orttraining/orttraining/core/optimizer/memory_optimizer/memory_insight.cc

```diff
@@ -0,0 +1,432 @@
+      continue;
+    }
+
+    bool can_compromise_stashed_activation = false;
+    std::shared_ptr<NodeRecomputePlan> recompute_plan =
+        CheckNodeForRecompute(*p_node,
+                              probe_level,
+                              fw_op_output_arg_used_map,
+                              node_index_to_its_order_in_topological_sort_map,
+                              candidate_output_args_map,
+                              logger, false,
+                              can_compromise_stashed_activation);
+    if (recompute_plan != nullptr) {
+      memory_opt_planner.AddNodeOptimizationPlan(p_node, recompute_plan);
+    }
+
+    if (can_compromise_stashed_activation) {
+      LOGS(logger, VERBOSE) << "Searching Node " << p_node->Name() << "(" << p_node->OpType()
+                            << ") for compromised recompute";
+      // If the subgraph recompute can save memory by comprising the assumption - recompute graphs' input must exist
+      // during backward pass, then we can consider to recomute them.
+      recompute_plan = CheckNodeForRecompute(*p_node, probe_level, fw_op_output_arg_used_map,
+                                             node_index_to_its_order_in_topological_sort_map,
+                                             candidate_output_args_map,
+                                             logger, true,
+                                             can_compromise_stashed_activation);
+      if (recompute_plan != nullptr) {
+        memory_opt_planner.AddNodeOptimizationPlan(p_node, recompute_plan);
+      }
+    }
+  }
+
+  return Status::OK();
+}
+
+void GetMemoryRecordsGroupedByNodeClusterId(const MemoryOptimizationPlanner& memory_opt_planner,
+                                            std::vector<std::pair<std::string, MemoryRecord>>& generated_records,
+                                            const NodeToClusterApplyContextMap& node_to_apply_contexts_map) {
+  // Group by node cluster id, generate memory record.
+  InlinedHashMap<std::string, MemoryRecord> records;
+  const auto& node_to_optimization_plan_map = memory_opt_planner.GetNodeToOptimizationPlanMap();
+  for (const auto& node_to_optimization_plan : node_to_optimization_plan_map) {
+    const auto& node = node_to_optimization_plan.first;
+    const auto& node_plans = node_to_optimization_plan.second;
+    const std::string node_cluster_id = memory_opt_planner.GenerateNodeClusterId(node);
+
+    bool already_exist = records.find(node_cluster_id) != records.end();
+
+    auto& record = records[node_cluster_id];
+    record.freq++;
+
+    // Collect more information for display.
+    for (auto& plan : node_plans) {
+      // Same cluster id, plans might still have different reuse_buffer pattern, so we need to collect all of them.
+      if (plan->reuse_buffers.size() > 0) {
+        const InlinedVector<size_t> output_indices = plan->GetActivationOutputIndices();
```

Comment:
>InlinedVector<size_t> output_indice [](http://example.com/codeflow?start=14&length=35)

This code does not seem to need a copy of the vector?

Consider returning a gsl::span per coding standards.